File: src/cli/commands/task.ts
import { Command } from 'commander'

import { getDb } from '../../db/index.js'
import { createTaskRepo, type TaskFilters } from '../../db/repositories/task.repo.js'
import {
  type ColumnDef,
  type FormatOptions,
  formatOutput,
  resolveFormat,
} from '../../formatters/index.js'
import { NotFoundError } from '../../models/errors.js'
import { PRIORITIES, type Priority } from '../../models/types.js'

const TASK_COLUMNS: ColumnDef[] = [
  { key: 'id', header: 'ID' },
  { key: 'title', header: 'Title' },
  { key: 'due_at', header: 'Due' },
  { key: 'priority', header: 'Priority' },
  { key: 'completed', header: 'Done' },
  { key: 'person_id', header: 'Person' },
]

interface GlobalOpts {
  format?: string
  quiet?: boolean
  db?: string
  color?: boolean
}

function getOpts(cmd: Command): GlobalOpts {
  let root = cmd
  while (root.parent) root = root.parent
  return root.opts()
}

function buildFormatOpts(globalOpts: GlobalOpts, columns?: ColumnDef[]): FormatOptions {
  const opts: FormatOptions = {}
  if (globalOpts.quiet !== undefined) opts.quiet = globalOpts.quiet
  if (globalOpts.color === false) opts.noColor = true
  if (columns) opts.columns = columns
  return opts
}

export function registerTaskCommands(program: Command): void {
  const task = program.command('task').description('Manage tasks & follow-ups')

  task
    .command('add')
    .description('Add a new task')
    .argument('<title>', 'Task title')
    .option('--person <id>', 'Person ID')
    .option('--deal <id>', 'Deal ID')
    .option('--due <date>', 'Due date')
    .option('--priority <level>', `Priority: ${PRIORITIES.join(', ')}`, 'normal')
    .option('--description <text>', 'Description')
    .action((title: string, opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const repo = createTaskRepo(db)

      const task = repo.create({
        title,
        description: (opts['description'] as string | undefined) ?? null,
        due_at: (opts['due'] as string | undefined) ?? null,
        priority: opts['priority'] as Priority | undefined,
        person_id: opts['person'] ? Number(opts['person']) : null,
        deal_id: opts['deal'] ? Number(opts['deal']) : null,
      })

      const format = resolveFormat(globalOpts.format)
      process.stdout.write(
        formatOutput([task], format, buildFormatOpts(globalOpts, TASK_COLUMNS)) + '\n',
      )
      if (!globalOpts.quiet) {
        process.stderr.write(`Created task #${String(task.id)}\n`)
      }
    })

  task
    .command('list')
    .description('List tasks')
    .option('--person <id>', 'Filter by person ID')
    .option('--deal <id>', 'Filter by deal ID')
    .option('--priority <level>', 'Filter by priority')
    .option('--overdue', 'Show only overdue tasks')
    .option('--done', 'Show completed tasks')
    .option('--pending', 'Show only pending tasks')
    .option('--sort <field>', 'Sort by: due, priority, created')
    .option('--limit <n>', 'Max results', '100')
    .action((opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const repo = createTaskRepo(db)

      const filters: TaskFilters = { limit: Number(opts['limit']) }
      if (opts['person']) filters.personId = Number(opts['person'])
      if (opts['deal']) filters.dealId = Number(opts['deal'])
      if (opts['priority']) filters.priority = opts['priority'] as Priority
      if (opts['overdue']) filters.overdue = true
      if (opts['done']) filters.completed = true
      if (opts['pending']) filters.completed = false
      if (opts['sort']) filters.sort = opts['sort'] as 'due' | 'priority' | 'created'

      const tasks = repo.findAll(filters)

      const format = resolveFormat(globalOpts.format)
      process.stdout.write(
        formatOutput(tasks, format, buildFormatOpts(globalOpts, TASK_COLUMNS)) + '\n',
      )
    })

  task
    .command('show')
    .description('Show task details')
    .argument('<id>', 'Task ID')
    .action((idStr: string, _opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const repo = createTaskRepo(db)
      const id = Number(idStr)

      const task = repo.findById(id)
      if (!task) throw new NotFoundError('task', id)

      const format = resolveFormat(globalOpts.format)
      process.stdout.write(formatOutput([task], format, buildFormatOpts(globalOpts)) + '\n')
    })

  task
    .command('edit')
    .description('Edit a task')
    .argument('<id>', 'Task ID')
    .option('--title <title>', 'Task title')
    .option('--description <text>', 'Description')
    .option('--due <date>', 'Due date')
    .option('--priority <level>', `Priority: ${PRIORITIES.join(', ')}`)
    .option('--person <id>', 'Person ID')
    .option('--deal <id>', 'Deal ID')
    .action((idStr: string, opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const repo = createTaskRepo(db)
      const id = Number(idStr)

      const existing = repo.findById(id)
      if (!existing) throw new NotFoundError('task', id)

      const update: Record<string, unknown> = {}
      if (opts['title'] !== undefined) update['title'] = opts['title']
      if (opts['description'] !== undefined) update['description'] = opts['description']
      if (opts['due'] !== undefined) update['due_at'] = opts['due']
      if (opts['priority'] !== undefined) update['priority'] = opts['priority']
      if (opts['person'] !== undefined) update['person_id'] = Number(opts['person'])
      if (opts['deal'] !== undefined) update['deal_id'] = Number(opts['deal'])

      const task = repo.update(id, update)
      if (!task) throw new NotFoundError('task', id)

      const format = resolveFormat(globalOpts.format)
      process.stdout.write(
        formatOutput([task], format, buildFormatOpts(globalOpts, TASK_COLUMNS)) + '\n',
      )
      if (!globalOpts.quiet) {
        process.stderr.write(`Updated task #${String(id)}\n`)
      }
    })

  task
    .command('done')
    .description('Mark a task as completed')
    .argument('<id>', 'Task ID')
    .action((idStr: string, _opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const repo = createTaskRepo(db)
      const id = Number(idStr)

      const task = repo.complete(id)
      if (!task) throw new NotFoundError('task', id)

      if (!globalOpts.quiet) {
        process.stderr.write(`Completed task #${String(id)}\n`)
      }
    })

  task
    .command('delete')
    .description('Delete a task (soft-delete)')
    .argument('<id>', 'Task ID')
    .action((idStr: string, _opts: Record<string, unknown>, cmd: Command) => {
      const globalOpts = getOpts(cmd)
      const db = getDb(globalOpts.db)
      const repo = createTaskRepo(db)
      const id = Number(idStr)

      const archived = repo.archive(id)
      if (!archived) throw new NotFoundError('task', id)

      if (!globalOpts.quiet) {
        process.stderr.write(`Archived task #${String(id)}\n`)
      }
    })
}

File: internal/cli/task.go
package cli

import (
	"fmt"
	"os"
	"strings"

	"github.com/jdanielnd/crm-cli/internal/db/repo"
	"github.com/jdanielnd/crm-cli/internal/format"
	"github.com/jdanielnd/crm-cli/internal/model"
	"github.com/spf13/cobra"
)

var taskColumns = []format.ColumnDef{
	{Header: "ID", Field: "id"},
	{Header: "Title", Field: "title"},
	{Header: "Due", Field: "due_at"},
	{Header: "Priority", Field: "priority"},
	{Header: "Done", Field: "completed"},
	{Header: "Person", Field: "person_id"},
}

func taskToMap(t *model.Task) map[string]any {
	m := map[string]any{
		"id":         t.ID,
		"uuid":       t.UUID,
		"title":      t.Title,
		"priority":   t.Priority,
		"completed":  t.Completed,
		"created_at": t.CreatedAt,
		"updated_at": t.UpdatedAt,
	}
	if t.Description != nil {
		m["description"] = *t.Description
	}
	if t.PersonID != nil {
		m["person_id"] = *t.PersonID
	}
	if t.DealID != nil {
		m["deal_id"] = *t.DealID
	}
	if t.DueAt != nil {
		m["due_at"] = *t.DueAt
	}
	if t.CompletedAt != nil {
		m["completed_at"] = *t.CompletedAt
	}
	return m
}

func tasksToMaps(tasks []*model.Task) []map[string]any {
	result := make([]map[string]any, len(tasks))
	for i, t := range tasks {
		result[i] = taskToMap(t)
	}
	return result
}

func registerTaskCommands(rootCmd *cobra.Command) {
	taskCmd := &cobra.Command{
		Use:   "task",
		Short: "Manage tasks and follow-ups",
	}

	taskCmd.AddCommand(taskAddCmd())
	taskCmd.AddCommand(taskListCmd())
	taskCmd.AddCommand(taskShowCmd())
	taskCmd.AddCommand(taskEditCmd())
	taskCmd.AddCommand(taskDoneCmd())
	taskCmd.AddCommand(taskDeleteCmd())

	rootCmd.AddCommand(taskCmd)
}

func taskAddCmd() *cobra.Command {
	var description, due, priority string
	var personID, dealID int64

	cmd := &cobra.Command{
		Use:   "add <title>",
		Short: "Add a new task",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			input := model.CreateTaskInput{
				Title:       args[0],
				Description: nilIfEmpty(description),
				Priority:    priority,
				DueAt:       nilIfEmpty(due),
			}
			if cmd.Flags().Changed("person") {
				input.PersonID = &personID
			}
			if cmd.Flags().Changed("deal") {
				input.DealID = &dealID
			}

			r := repo.NewTaskRepo(db)
			task, err := r.Create(cmd.Context(), input)
			if err != nil {
				return err
			}

			data := []map[string]any{taskToMap(task)}
			return format.Output(os.Stdout, resolveFormat(), data, taskColumns, flagQuiet)
		},
	}

	cmd.Flags().StringVar(&description, "description", "", "task description")
	cmd.Flags().StringVar(&due, "due", "", "due date (ISO 8601)")
	cmd.Flags().StringVar(&priority, "priority", "medium", "priority: "+strings.Join(model.Priorities, ", "))
	cmd.Flags().Int64Var(&personID, "person", 0, "associated person ID")
	cmd.Flags().Int64Var(&dealID, "deal", 0, "associated deal ID")

	return cmd
}

func taskListCmd() *cobra.Command {
	var personID, dealID int64
	var overdue, all bool
	var limit int

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List tasks",
		RunE: func(cmd *cobra.Command, args []string) error {
			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			filters := model.TaskFilters{
				Limit:            limit,
				Overdue:          overdue,
				IncludeCompleted: all,
			}
			if cmd.Flags().Changed("person") {
				filters.PersonID = &personID
			}
			if cmd.Flags().Changed("deal") {
				filters.DealID = &dealID
			}

			r := repo.NewTaskRepo(db)
			tasks, err := r.FindAll(cmd.Context(), filters)
			if err != nil {
				return err
			}

			return format.Output(os.Stdout, resolveFormat(), tasksToMaps(tasks), taskColumns, flagQuiet)
		},
	}

	cmd.Flags().Int64Var(&personID, "person", 0, "filter by person ID")
	cmd.Flags().Int64Var(&dealID, "deal", 0, "filter by deal ID")
	cmd.Flags().BoolVar(&overdue, "overdue", false, "show only overdue tasks")
	cmd.Flags().BoolVar(&all, "all", false, "include completed tasks")
	cmd.Flags().IntVar(&limit, "limit", 0, "max results")

	return cmd
}

func taskShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show task details",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			id, err := parseEntityID(args[0], "task")
			if err != nil {
				return err
			}

			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			r := repo.NewTaskRepo(db)
			task, err := r.FindByID(cmd.Context(), id)
			if err != nil {
				return err
			}

			data := []map[string]any{taskToMap(task)}
			return format.Output(os.Stdout, resolveFormat(), data, taskColumns, flagQuiet)
		},
	}
}

func taskEditCmd() *cobra.Command {
	var title, description, due, priority string
	var personID, dealID int64

	cmd := &cobra.Command{
		Use:   "edit <id>",
		Short: "Edit a task",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			id, err := parseEntityID(args[0], "task")
			if err != nil {
				return err
			}

			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			input := model.UpdateTaskInput{}
			if cmd.Flags().Changed("title") {
				input.Title = &title
			}
			if cmd.Flags().Changed("description") {
				input.Description = &description
			}
			if cmd.Flags().Changed("due") {
				input.DueAt = &due
			}
			if cmd.Flags().Changed("priority") {
				input.Priority = &priority
			}
			if cmd.Flags().Changed("person") {
				input.PersonID = &personID
			}
			if cmd.Flags().Changed("deal") {
				input.DealID = &dealID
			}

			r := repo.NewTaskRepo(db)
			task, err := r.Update(cmd.Context(), id, input)
			if err != nil {
				return err
			}

			data := []map[string]any{taskToMap(task)}
			return format.Output(os.Stdout, resolveFormat(), data, taskColumns, flagQuiet)
		},
	}

	cmd.Flags().StringVar(&title, "title", "", "task title")
	cmd.Flags().StringVar(&description, "description", "", "task description")
	cmd.Flags().StringVar(&due, "due", "", "due date (ISO 8601)")
	cmd.Flags().StringVar(&priority, "priority", "", "priority: "+strings.Join(model.Priorities, ", "))
	cmd.Flags().Int64Var(&personID, "person", 0, "associated person ID")
	cmd.Flags().Int64Var(&dealID, "deal", 0, "associated deal ID")

	return cmd
}

func taskDoneCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "done <id>",
		Short: "Mark a task as completed",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			id, err := parseEntityID(args[0], "task")
			if err != nil {
				return err
			}

			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			r := repo.NewTaskRepo(db)
			task, err := r.Complete(cmd.Context(), id)
			if err != nil {
				return err
			}

			fmt.Fprintf(os.Stderr, "Completed task #%d: %s\n", task.ID, task.Title)
			return nil
		},
	}
}

func taskDeleteCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete <id>",
		Short: "Delete a task (soft-delete)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			id, err := parseEntityID(args[0], "task")
			if err != nil {
				return err
			}

			db, err := openDB()
			if err != nil {
				return err
			}
			defer db.Close()

			r := repo.NewTaskRepo(db)
			if err := r.Archive(cmd.Context(), id); err != nil {
				return err
			}

			fmt.Fprintf(os.Stderr, "Deleted task #%d\n", id)
			return nil
		},
	}
}
